service/matching/task_queue_manager.go

```diff
@@ -783,40 +813,42 @@ func (c *taskQueueManagerImpl) userDataFetchSource() (string, error) {
 func (c *taskQueueManagerImpl) fetchUserData(ctx context.Context) error {
 	ctx = c.callerInfoContext(ctx)
 
-	if !c.config.LoadUserData() {
-		// if disabled, mark ready now
-		c.SetUserDataInitialFetch(nil)
-		return nil
-	}
 	if c.managesSpecificVersionSet() {
 		// tqm for specific version set doesn't have its own user data
-		c.SetUserDataInitialFetch(nil)
+		c.SetUserDataStatus(userDataSpecificVersion, nil)
 		return nil
 	}
-	if c.db.DbStoresUserData() {
-		// root workflow partition "owns" user data, read it from db
-		err := c.db.loadUserData(ctx)
-		c.SetUserDataInitialFetch(err)
-		return err
-	}
 
 	// otherwise fetch from parent partition
 
 	fetchSource, err := c.userDataFetchSource()
 	if err != nil {
 		if err == errMissingNormalQueueName { // nolint:goerr113
-			// pretend we have no user data
-			c.SetUserDataInitialFetch(nil)
+			// pretend we have no user data. this is a sticky queue so the only effect is that we can't
+			// kick off versioned pollers.
+			c.SetUserDataStatus(userDataEnabled, nil)
 		}
 		return err
 	}
 
-	firstCall := true
+	// hasFetchedUserData is true if we have gotten a successful reply to GetTaskQueueUserData.
+	// It's used to control whether we do a long poll or a simple get.
```

Comment:
I think `long poll` is a bit confusing here, especially since it's a one-of-a-kind concept in server. I think let's just elaborate on what exactly this is. Like, a long poll on a task queue makes sense because you're waiting for a task, but a long poll on something that seems like static data (metadata of a task queue), doesn't immediatley make sense. I'd explain that we're waiting for any updates to the user data, and that we do that by sending a request with our current version that blocks until the user data on the server is updated to a higher version.

Reply:
"long poll" is a general concept for waiting for any sort of new data, and this does fit the pattern. It is used elsewhere in the server and referred to by that name, e.g. getting history and waiting for new events (https://github.com/temporalio/temporal/blob/83faadf3cedcdef689b88d911cfe9d22d36fc9d5/service/history/api/get_workflow_util.go#L95 and elsewhere)